### ISSUE TITLE
Intermittent missing labels after text classification

### ISSUE THREAD

**Issue by pjtinker** [flairNLP/flair#559]
I have trained a text classification model using Glove and base uncased BERT embeddings.  However, when used to classify unseen data, the model will occasionally return an empty list for a given sample.  

For instance,
```
clf = TextClassifier.load_from_file('./bert_embeddings/best-model.pt')

text = "i do not think so but i do not know very much on the topic i do not know what data you should have"
token = ''
for word in text.split():
    token += word + ' '
    sentence = Sentence(token)
    clf.predict(sentence)
    print("{} \n {}".format(token, sentence.labels))
```
yields: 
```
i
 [0 (0.9997053742408752)]
i do
 [0 (0.9999978542327881)]
i do not
 [0 (0.9999978542327881)]
i do not think
 [0 (0.9999986886978149)]
i do not think so
 [0 (0.9999988079071045)]
i do not think so but
 [0 (0.9999992847442627)]
i do not think so but i
 [0 (0.9999997615814209)]
i do not think so but i do
 [0 (0.9999997615814209)]
i do not think so but i do not
 [0 (0.9999997615814209)]
i do not think so but i do not know
 [0 (0.9999995231628418)]
i do not think so but i do not know very
 [0 (0.9999991655349731)]
i do not think so but i do not know very much
 [0 (0.999996542930603)]
i do not think so but i do not know very much on
 [0 (0.9999967813491821)]
i do not think so but i do not know very much on the
 [0 (0.9999806880950928)]
i do not think so but i do not know very much on the topic
 []
i do not think so but i do not know very much on the topic i
 [0 (0.9809603691101074)]
i do not think so but i do not know very much on the topic i do
 [0 (0.9999698400497437)]
i do not think so but i do not know very much on the topic i do not
 [0 (0.9990166425704956)]
i do not think so but i do not know very much on the topic i do not know
 [0 (0.9852306246757507)]
i do not think so but i do not know very much on the topic i do not know what
 [0 (0.9929770827293396)]
i do not think so but i do not know very much on the topic i do not know what data
 [0 (0.96560138463974)]
i do not think so but i do not know very much on the topic i do not know what data you
 [0 (0.8830254673957825)]
i do not think so but i do not know very much on the topic i do not know what data you should
 []
i do not think so but i do not know very much on the topic i do not know what data you should have
 []
```

The model was trained following the instructions given in the tutorial.  
**Environment:**
 - CentOS 7
 - Flair 0.4.1
 - Python 3.6

If anyone has any suggestions or insight, it would be greatly appreciated!

**Update**
A model trained using Glove, mix-forward, and mix-backward Flair embeddings has the same issue.

**Comment by alanakbik** [comment]
Hello @pjtinker that seems strange - could you give a bit more info about the classification task: Is this a binary classification task? Did you train with multi_label = False?

**Comment by pjtinker** [comment]
Hi @alanakbik.  This is multi-class, with four possible labels.  I trained with `multi_label = True`

Here is the code block used to generate word embeddings and train ripped straight from the tutorial except for the use of DocumentRNNEmbeddings():
```
word_embeddings = [WordEmbeddings('glove'),

                   # comment in flair embeddings for state-of-the-art results 
                   BertEmbeddings('bert-base-uncased')  
		   #FlairEmbeddings('mix-forward'),
                   #FlairEmbeddings('mix-backward'),
                   ]

# 4. initialize document embedding by passing list of word embeddings
document_embeddings: DocumentRNNEmbeddings = DocumentRNNEmbeddings(word_embeddings,
								     rnn_type='LSTM',
                                                                     hidden_size=512,
                                                                     reproject_words=True,
                                                                     reproject_words_dimension=256,
                                                                     )

# 5. create the text classifier
classifier = TextClassifier(document_embeddings, label_dictionary=label_dict, multi_label=True)

# 6. initialize the text classifier trainer
trainer = ModelTrainer(classifier, corpus)
```

**Comment by alanakbik** [comment]
In this case, the behavior might be normal. Multi-labels classification can return multiple labels, and none at all. So an empty list may mean that the classifier is unable to find a fitting label for the sentence. 

But it is strange since the sentences without labels look almost the same as the ones with labels but maybe the classifier overfits to some signal that then causes these misclassifications?

**Comment by amit8121** [comment]
@pjtinker @alanakbik 

I have the same issue with  `multi label classification` task. I used `DocumentRNNEmbeddings` with `flair-news-forward` and `bert-base-uncased` as word embeddings. 

If it's a common issue for `multi label classification` task, it could be pretty dangerous to use this model in production right ? What are the best ways that we can use to counter situations where it returns empty list ?

Alan is right as per over-fitting of the data. I trained for only `15 epochs` yet my model over fits a great deal.

**The Code**
```

bert_base_embedding = BertEmbeddings('bert-base-uncased', pooling_operation = 'mean')
flair_embedding_forward = FlairEmbeddings('news-forward-fast', chars_per_chunk=256)

word_embeddings = [bert_base_embedding, flair_embedding_forward]

document_embeddings: DocumentRNNEmbeddings = DocumentRNNEmbeddings(word_embeddings,
                                                                     hidden_size=512, 
                                                                     rnn_layers = 2,
                                                                     reproject_words=True,
                                                                     reproject_words_dimension=128,
                                                                     bidirectional = True,
                                                                     dropout = 0.35 , 
                                                                     rnn_type = 'GRU'
                                                                     )




classifier = TextClassifier(document_embeddings, label_dictionary=label_dict, multi_label = True)


trainer = ModelTrainer(classifier, corpus,
                      optimizer = Adam)


trainer.train('resources/taggers/ag_news',
               EvaluationMetric.MICRO_F1_SCORE,
              eval_mini_batch_size = 20,
#                learning_rate = 1e-1,
              learning_rate=1e-3,
              amsgrad = True,
              weight_decay = 5e-5,
              mini_batch_size=32,
              anneal_factor=0.4,
              patience=3,
              max_epochs=15)


```


**The Results** 

![Over fitting](https://user-images.githubusercontent.com/23751321/61485984-4d8f5a00-a957-11e9-89cd-8ede0f6e5675.PNG)


My training loss eventually reaches to `0.015` but test and validation accuracies never improves, a classic case of over-fitting. No matter how much ever I increase the `dropout probability` and other related hyper-paramters, the over-fitting issue stays the same. It might be the case because of the sere size of the data as well. But really keen to know how to tackle the case where there's no label returned.

TIA for the response !

**Comment by alanakbik** [comment]
@amit8121 you could try reducing the capacity of the RNN, which is currently very powerful. Try going for one layer and maybe only 256 hidden states to counter overfitting. 

If you use the current master branch there are ways for specifying the classification threshold in a multi-label classification task: 

```python
# set a lower classification threshold if you want
classifier.multi_label_threshold = 0.1
```

**Comment by yzhaoinuw** [comment]
> Hello @pjtinker that seems strange - could you give a bit more info about the classification task: Is this a binary classification task? Did you train with multi_label = False?

Hi @alanakbik if it is a single label and a binary classification, is CrossEntropyLoss equivalent to BCEWithLogitsLoss? I saw that CrossEntropyLoss is used in the TextClassifier class when it is a single label classification. By the way, it's not necessary to specify multi_label=False right, since TextClassifier will count the number of class, or it's better to specify?
